pkg/controller/directvolumemigration/task.go

```diff
@@ -0,0 +1,409 @@
+	all uint8
+	// Step included when ANY flag evaluates true.
+	any uint8
+}
+
+type Itinerary struct {
+	Name  string
+	Steps []Step
+}
+
+// Get a progress report.
+// Returns: phase, n, total.
+func (r Itinerary) progressReport(phase string) (string, int, int) {
+	n := 0
+	total := len(r.Steps)
+	for i, step := range r.Steps {
+		if step.phase == phase {
+			n = i + 1
+			break
+		}
+	}
+
+	return phase, n, total
+}
+
+var PVCItinerary = Itinerary{
+	Name: "PVC",
+	Steps: []Step{
+		{phase: Created},
+		{phase: Started},
+		{phase: Prepare},
+		{phase: CreateDestinationNamespaces},
+		{phase: DestinationNamespacesCreated},
+		{phase: CreateDestinationPVCs},
+		{phase: DestinationPVCsCreated},
+		{phase: CreateRsyncRoute},
+		{phase: CreateRsyncConfig},
+		{phase: CreateStunnelConfig},
+		{phase: CreateRsyncTransferPods},
+		{phase: WaitForRsyncTransferPodsRunning},
+		{phase: CreateStunnelClientPods},
+		{phase: WaitForStunnelClientPodsRunning},
+		{phase: CreateRsyncClientPods},
+		{phase: WaitForRsyncClientPodsCompleted},
+		{phase: DeleteRsyncResources},
+		{phase: Completed},
+	},
+}
+
+var FailedItinerary = Itinerary{
+	Name: "Failed",
+	Steps: []Step{
+		{phase: MigrationFailed},
+		{phase: Completed},
+	},
+}
+
+// A task that provides the complete migration workflow.
+// Log - A controller's logger.
+// Client - A controller's (local) client.
+// Owner - A DirectVolumeMigration resource.
+// Phase - The task phase.
+// Requeue - The requeueAfter duration. 0 indicates no requeue.
+// Itinerary - The phase itinerary.
+// Errors - Migration errors.
+// Failed - Task phase has failed.
+type Task struct {
+	Log         logr.Logger
+	Client      k8sclient.Client
+	Owner       *migapi.DirectVolumeMigration
+	SSHKeys     *sshKeys
+	RsyncRoutes map[string]string
+	Phase       string
+	Requeue     time.Duration
+	Itinerary   Itinerary
+	Errors      []string
+}
+
+type sshKeys struct {
+	PublicKey  *rsa.PublicKey
+	PrivateKey *rsa.PrivateKey
+}
+
+func (t *Task) init() error {
+	t.RsyncRoutes = make(map[string]string)
+	t.Requeue = FastReQ
+	if t.failed() {
+		t.Itinerary = FailedItinerary
+	} else {
+		t.Itinerary = PVCItinerary
+	}
+	return nil
+}
+
+func (t *Task) Run() error {
+	t.Log.Info("[RUN]", "phase", t.Phase)
+
+	err := t.init()
+	if err != nil {
+		return err
+	}
+
+	// Run the current phase.
+	switch t.Phase {
+	case Created, Started:
+		if err = t.next(); err != nil {
+			return liberr.Wrap(err)
+		}
+	case Prepare:
+		if err = t.next(); err != nil {
+			return liberr.Wrap(err)
+		}
+	case CreateDestinationNamespaces:
+		// Create all of the namespaces the migrated PVCs are in are created on the
+		// destination
+		err := t.ensureDestinationNamespaces()
+		if err != nil {
+			liberr.Wrap(err)
+		}
+		t.Requeue = NoReQ
+		if err = t.next(); err != nil {
+			return liberr.Wrap(err)
+		}
+	case DestinationNamespacesCreated:
+		// Ensure the namespaces are created
+		// TODO: bad func name
+		err := t.getDestinationNamespaces()
```

Comment:
I doubt how useful this is?

Once the namespace object is created, I am not sure what options you have on the object status to check. I would assume that if the create call did not error in the above phase unless some agent explicitly calls delete on the namespace, it will be created

Reply:
hrm actually

```
A namespace can be in one of two phases:

- Active the namespace is in use
- Terminating the namespace is being deleted, and can not be used for new objects
```

You can check for `Active` status of the namespace, so might be useful, still doubtful

Reply:
Yeah I agree... part of the reason why this function just returns nil. I didn't see much value. I'm good to delete this.